### PR TITLE
Add break period display and warning arrows

### PR DESCRIPTION
**Pull request by EVAST9919**

### CODE REVIEW

File: osu.Game/Screens/Play/Player.cs [hunk 1]
```diff
@@ -194,6 +178,22 @@ private void load(AudioManager audio, BeatmapDatabase beatmaps, OsuConfigManager
...
                             Origin = Anchor.Centre
                         },
+                        new SkipButton(firstObjectTime) { AudioClock = decoupledClock },
+                        sectionTrackOverlay = new SectionTrackOverlay(firstObjectTime)
+                        {
+                            Breaks = Beatmap.Beatmap.Breaks,
+                            AudioClock = decoupledClock,
+                            BreakIn = () =>
+                            {
+                                Background?.FadeTo(1, fade_duration);
```
Comment (ColdVolcano) [inline osu.Game/Screens/Play/Player.cs]: I'm almost sure dimLevel should also play a role here

File: osu.Game/Screens/Play/SectionTrackOverlay.cs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            private const int appear_duration = 140;
+            private const int margin_vertical = 120;
+            private const int margin_horizontal = 90;
+
+            private readonly Container content;
+
+            public ArrowsOverlay()
+            {
+                RelativeSizeAxes = Axes.Both;
+                Add(content = new Container
```
Comment (ColdVolcano) [inline osu.Game/Screens/Play/SectionTrackOverlay.cs]: Is there a need for an extra level here?

**Comment by peppy** [comment]
what does "basic section tracking" even men

File: osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+            private readonly Container content;
+
+            public ArrowsOverlay()
+            {
+                RelativeSizeAxes = Axes.Both;
+                Add(content = new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Alpha = 0,
+                    Children = new Drawable[]
```
Comment (ColdVolcano) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: Why not set children directly instead of adding a container with said children

Comment (EVAST9919) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: We are doing the same thing in HudOverlay. I guess, we don't wanna hide the container itself, only the content inside this container.

File: osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                    }
+                    else
+                    {
+                        resultIcon.Icon = FontAwesome.fa_check;
+                        samplePass.Play();
+                    }
+
+                    resultIcon.FadeTo(1);
+                    Delay(100);
+                    Schedule(() => resultIcon.FadeTo(0));
```
Comment (peppy) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: just use delay on the `resultIcon`  directly and don't use `Schedule`

**Comment by peppy** [comment]
No TestCase? :(

File: osu.Game/Screens/Play/Player.cs [hunk 5]
```diff
@@ -155,7 +164,38 @@ private void load(AudioManager audio, OsuConfigManager config, OsuGame osu)
                     },
                     Children = new Drawable[]
                     {
-                        new SkipButton(firstObjectTime) { AudioClock = decoupledClock },
+                        letterboxInBreaksContainer = new Container
```
Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: Please move this implementation to a new container; you're polluting `Player` again.

Comment (peppy) [inline osu.Game/Screens/Play/Player.cs]: Also make it IStateful and move the hide/show transform logic inside it.

File: osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Timing;
+using osu.Game.Beatmaps.Timing;
+using osu.Game.Graphics;
+using System;
+using System.Collections.Generic;
+
+namespace osu.Game.Screens.Play
+{
+    public class BreakPeriodsTrackOverlay : Container
```
Comment (peppy) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: `BreakOverlay` sounds fine

File: osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    {
+        private const double fade_duration = BreakPeriod.MIN_BREAK_DURATION_FOR_EFFECT / 2;
+        private const int arrows_appear_offset = 1000;
+
+        private readonly BindableDouble healthBindable = new BindableDouble
```
Comment (peppy) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: no need for bindable suffix

File: osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        {
+            samplePass = audio.Sample.Get(@"Gameplay/sectionpass");
+            sampleFail = audio.Sample.Get(@"Gameplay/sectionfail");
+        }
+
+        public void BindHealth(BindableDouble health) => healthBindable.BindTo(health);
+
+        protected override void Update()
+        {
+            if (breaks == null) return;
```
Comment (peppy) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: this whole method is what?

Comment (peppy) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: you can't just overload update to do stuff like this. it's shocking for readability and extensibility.

**Review by peppy — CHANGES_REQUESTED** [review]
needs a lot of refactoring

File: osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs (hunk 7 above)
Comment (EVAST9919) [inline osu.Game/Screens/Play/BreakPeriodsTrackOverlay.cs]: That's because we already have `health` variable

**Comment by EVAST9919** [comment]
Will be refactored and reimplemened in future pr's

**Pull request closed without merging**